packages/flutter/lib/src/material/carousel.dart

```diff
@@ -0,0 +1,744 @@
+  /// Similarly, if [scrollDirection] is [Axis.vertical], then the page view
+  /// scrolls from top to bottom when [reverse] is false and from bottom to top
+  /// when [reverse] is true.
+  ///
+  /// Defaults to false.
+  final bool reverse;
+
+  /// Called when one of the [children] is tapped.
+  final ValueChanged<int>? onTap;
+
+  /// The extent the children are forced to have in the main axis.
+  final double? itemExtent;
+
+  /// The child widgets for carousel.
+  final List<Widget> children;
+
+  @override
+  State<Carousel> createState() => _CarouselState();
+}
+
+class _CarouselState extends State<Carousel> {
+  late double? itemExtent;
+  CarouselController? _internalController;
+  CarouselController get _controller => widget.controller ?? _internalController!;
+
+  @override
+  void initState() {
+    if (widget.controller == null) {
+      _internalController = CarouselController();
+    }
+    _controller._attach(this);
+    super.initState();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    itemExtent = getItemExtent();
+  }
+
+  @override
+  void didUpdateWidget(covariant Carousel oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller?._detach(this);
+      if (widget.controller != null) {
+        _internalController?._detach(this);
+        _internalController = null;
+        widget.controller?._attach(this);
+      }
+    }
+    if (widget.itemExtent != oldWidget.itemExtent) {
+      itemExtent = getItemExtent();
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller._detach(this);
+    if (widget.controller == null) {
+      _controller.dispose();
+    }
+    super.dispose();
+  }
+
+  AxisDirection _getDirection(BuildContext context) {
+    switch (widget.scrollDirection) {
+      case Axis.horizontal:
+        assert(debugCheckHasDirectionality(context));
+        final TextDirection textDirection = Directionality.of(context);
+        final AxisDirection axisDirection = textDirectionToAxisDirection(textDirection);
+        return widget.reverse ? flipAxisDirection(axisDirection) : axisDirection;
+      case Axis.vertical:
+        return widget.reverse ? AxisDirection.up : AxisDirection.down;
+    }
+  }
+
+  double? getItemExtent() {
+    if (widget.itemExtent != null) {
+      final double screenExtent = switch(widget.scrollDirection) {
+        Axis.horizontal => MediaQuery.of(context).size.width,
+        Axis.vertical => MediaQuery.of(context).size.height,
+      };
+
+      return clampDouble(widget.itemExtent!, 0, screenExtent);
+    }
+    return null;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final AxisDirection axisDirection = _getDirection(context);
+    final ScrollPhysics physics = widget.itemSnapping
+      ? const CarouselScrollPhysics()
+      : ScrollConfiguration.of(context).getScrollPhysics(context);
```

Comment:
Do we have a `ScrollConfiguration.scrollPhysicsOf(context)` method? It feels a bit weird to have to use the context twice here. Do I file an issue?

Reply:
Done :)